aws/sdk/s3-benchmark/benchmark/src/multipart_get.rs

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::{Args, BoxError, BENCH_KEY};
+use aws_sdk_s3 as s3;
+use aws_smithy_http::byte_stream::ByteStream;
+use std::fmt;
+use std::path::Path;
+use std::sync::Arc;
+use tokio::sync::Semaphore;
+
+pub async fn get_object_multipart(
+    client: &s3::Client,
+    args: &Args,
+    path: &Path,
+) -> Result<(), BoxError> {
+    let mut part_count = (args.size_bytes / args.part_size_bytes + 1) as i64;
+    let mut size_of_last_part = (args.size_bytes % args.part_size_bytes) as i64;
+    if size_of_last_part == 0 {
+        size_of_last_part = args.part_size_bytes as i64;
+        part_count -= 1;
+    }
+
+    let ranges = (0..part_count).map(|i| {
+        if i == part_count - 1 {
+            let start = i * args.part_size_bytes as i64;
+            ContentRange::new(start, start + size_of_last_part - 1, size_of_last_part)
+        } else {
+            ContentRange::new(
+                i * args.part_size_bytes as i64,
+                (i + 1) * args.part_size_bytes as i64 - 1,
+                args.part_size_bytes as i64,
+            )
+        }
+    });
+
+    let semaphore = Arc::new(Semaphore::new(args.concurrency));
+    let mut tasks = Vec::new();
+    for range in ranges {
+        tasks.push(tokio::spawn(download_part(
```

Comment:
There's a slight problem here in that the tasks will run in a non-deterministic order, which means later parts of the file can start downloading first. Since the tasks are iterated in order later, and since the body is read upon awaiting the tasks, this could potentially lead to issues if it takes a long time to get to one of the later parts that started sooner than everything else.

Not sure if this is really a problem in practice, or just hypothetical. Also not really sure if there's an easy way to fix it.

Reply:
I'd fix it by:
1. Pushing the work items onto a channel (in order)
2. Have the tasks 1) Acquire semaphore, 2) pull item from channel

That keeps the work items weighted to the front of the file which means the early tasks finish first. 

This is likely to be a little faster in practice, but more importantly, it will result in lower (and bounded!) memory usage.

Reply:
Ah, good idea. Thanks for the suggestion!